### DISCUSSION TITLE
How to best expose Util methods to CC that require access to mod config options

### DISCUSSION THREAD

**Discussion by pupnewfster** [cc-tweaked/CC-Tweaked#947] [Q&A] [answered]
Hi, I was wondering if there is a decent way to expose custom "programs" to CC except that has access to the java side of code. I know that it is possible to add lua scripts to a datapack but I am not sure offhand of any way to make it reference the java side of things. In Mekanism internally all our energy values are in Joules, but given a lot of players probably are going to want to be converting to FE if possible, which I feel like would make the most sense as a "helper" program so that people can say do `joulesToFE` or `feToJoules`. The main reason that this would "need" to be accessing the java side of code is that the various conversion rates (including if they can even be converted) are config based so it would be good to be able to have it make use of the actual values rather than be hardcoded to what the default rates are.

In theory I could slightly tweak my [system](https://github.com/mekanism/Mekanism/blob/66c95bca302900aa7ca973f03584b59f7fde064f/src/main/java/mekanism/common/integration/computer/ComputerMethodMapper.java#L176-L232) that converts annotations into [multiple methods](https://github.com/mekanism/Mekanism/blob/v10.1/src/main/java/mekanism/common/integration/computer/SpecialComputerMethodWrapper.java) especially for returns, and make it so that say every `getEnergyStored` becomes `getEnergyStored`, `getEnergyStoredFE`, `getEnergyStoredEU` (once IC2 updates and we support it properly again), but it feels like that would be a bit overkill and also be rather messy for doing it for setters as well if they accept multiple parameters, and also be a bit confusing to the user.

**Comment by SquidDev** [comment, marked as answer]
Would [`ILuaAPI`](https://github.com/cc-tweaked/CC-Tweaked/blob/mc-1.16.x/src/main/java/dan200/computercraft/api/lua/ILuaAPI.java) do what you want? This allows you to register a global API, so people could do `mekanism.joulesToFE(123)` or something.

It's not a perfect solution - new globals are always gross - but I'm not sure of a better solution here.

**Reply by pupnewfster** [reply]
Yes from a glance at `ILuaAPI` I think that might work well as a solution to this, thanks!
